### PR TITLE
Add Types: support for InfiltrateFor* traits

### PR DESCRIPTION
**Pull request by MustaphaTR**
This is required for RA Classic mod as there are 2 infiltrators (Spy and Thief) that do different things in same building.

~~For TESTCASE i buffed British Spy to give a better sonar pulse, create power outage for double the time and steal 75% of the cash instead of 50%.~~

### CODE REVIEW

File: OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForCash.cs [hunk 1]
```diff
@@ -43,6 +46,10 @@ class InfiltrateForCash : INotifyInfiltrated
 
 		void INotifyInfiltrated.Infiltrated(Actor self, Actor infiltrator)
 		{
+			var infiltrates = infiltrator.Info.TraitInfo<InfiltratesInfo>();
```
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForCash.cs]: Can you please instead change the interface to pass the type through the notification instead?
```csharp
	public interface INotifyInfiltrated { void Infiltrated(Actor self, Actor infiltrator, HashSet<string> types); }
```

This follows the convention from our other types, and makes it easier to support multiple `Infiltrates` traits on an actor.

Comment (pchote) [reply OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForCash.cs]: You can then change `Infiltrate.cs` to use
```csharp
			foreach (var t in target.TraitsImplementing<INotifyInfiltrated>())
				t.Infiltrated(target, self, infiltrates.Info.Types);
```

**Review by pchote — DISMISSED** [review]

**Comment by MustaphaTR** [comment]
Updated.

**Review by GraionDilach — APPROVED** [review]
:+1:

**Comment by MustaphaTR** [comment]
Removed the testcase.

**Comment by pchote** [comment]
This makes a backwards incompatible change to the mod rules (specifying `Types` is required for the effects to work), so this really should have an upgrade rule warning about this.  But... the current UpgradeRules code doesn't cleanly support the way this warning should work (only warn once, if it encounters one of the affected traits) and so this can wait until the upgrade rule overhaul that makes it easy.

**Review by pchote — APPROVED** [review]
LGTM otherwise

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9484c34755626fa11d84c88f7bbed3f35a79ca83

**Pull request merged**